feat: static color coding for key column

danferns opened the pull request:
Hi! I just got the core of this feature working and I'm seeking feedback.

A small rectangle is drawn next to the Key label in the Library. The color of the rectangle is based on the Key of the track.

![image](https://github.com/mixxxdj/mixxx/assets/57069381/c4536a72-dcd4-47e6-a2e5-58f140b99f00)

---

I'm concerned about the method used for finding the track's key. Right now, I'm using `KeyUtils::guessKeyFromText` on the Key string, but I don't think that'll work if the user sets custom Key Notation. 

I want to use the value in [`ColumnCache::COLUMN_LIBRARYTABLE_KEY_ID`](https://github.com/danferns/mixxx/blob/7931a0756cd8fb00422806d685c7029530864f14/src/library/basetracktablemodel.cpp#L888). But I haven't been able to figure out how to access this column from the `QModelIndex& index` that is passed into `KeyDelegate::paintItem`. I'm not sure if the column is even available.

Is there a place where I can see how the index is structured?

---

Once complete, this will resolve #7810.

danferns committed:
add KeyDelegate

danferns committed:
add Default Key Color Palette

danferns committed:
use Key Color Palette to change Key Column background

danferns committed:
implement the design for Key Column colors

danferns committed:
style: increase right padding for key color rectangle

Swiftb0y commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+#include "library/tabledelegates/keydelegate.h"
+
+#include <qnamespace.h>
```
weird include, remove?
```suggestion
```

Swiftb0y commented on src/library/tabledelegates/keydelegate.h:
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include "library/tabledelegates/tableitemdelegate.h"
+
```
also `#include <QObject>`.

Swiftb0y commented:
> Once complete, this will resolve https://github.com/mixxxdj/mixxx/issues/7810.

I don't think so. That issue requests to color the key depending on how compatible it is to the currently playing track.

Swiftb0y commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+#include "util/color/predefinedcolorpalettes.h"
+#include "util/color/rgbcolor.h"
+
+KeyDelegate::KeyDelegate(QTableView* pTableView)
+        : TableItemDelegate(pTableView) {
+}
```
move into header please.

daschuer replied:
What's the reason for this request?

Swiftb0y replied:
potentially performance. constructors this trivial will likely always get inlined, but the compiler needs it in the header for that.

Swiftb0y commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+    int openKeyNumber = KeyUtils::keyToOpenKeyNumber(key);
+
+    if (openKeyNumber != 0) {
+        const auto rgbColor =
+                mixxx::PredefinedColorPalettes::kDefaultKeyColorPalette.at(
+                        openKeyNumber - 1); // Open Key numbers start from 1
```
check bounds as well.
```suggestion
        const auto& palette = mixxx::PredefinedColorPalettes::kDefaultKeyColorPalette;
        DEBUG_ASSERT(openKeyNumber <= palette.length());
        const auto rgbColor =
                palette.at(
                        openKeyNumber - 1); // Open Key numbers start from 1
```

Swiftb0y commented on src/util/color/predefinedcolorpalettes.cpp:
```diff
@@ -144,6 +144,21 @@ constexpr mixxx::RgbColor kVirtualDJTrackColorBlue(0x0000FF);
 
+// Default Mixxx Key Color Palette
+
+constexpr mixxx::RgbColor kMixxxKeyColor1(0xFC4949);
+constexpr mixxx::RgbColor kMixxxKeyColor2(0xFE642D);
+constexpr mixxx::RgbColor kMixxxKeyColor3(0xF98C27);
+constexpr mixxx::RgbColor kMixxxKeyColor4(0xFED600);
+constexpr mixxx::RgbColor kMixxxKeyColor5(0x99FE00);
+constexpr mixxx::RgbColor kMixxxKeyColor6(0x42FE3E);
+constexpr mixxx::RgbColor kMixxxKeyColor7(0x0AD58F);
+constexpr mixxx::RgbColor kMixxxKeyColor8(0x0AE7E7);
+constexpr mixxx::RgbColor kMixxxKeyColor9(0x04C9FE);
+constexpr mixxx::RgbColor kMixxxKeyColor10(0x3D8AFD);
+constexpr mixxx::RgbColor kMixxxKeyColor11(0xAC64FE);
+constexpr mixxx::RgbColor kMixxxKeyColor12(0xFD3FEA);
+
```
as a follow up, it would be nice if the palette was customizable via the `Color` Preferences. similar to track colors.

ronso0 replied:
Super nice to have, follow up²: set the colors in the notation table in Key preferences

Swiftb0y replied:
Yeah, Though I guess we'll have to decide on whether its better to put in the Color or the Key settings. The former would be more consistent IMO.

danferns replied:
Here's a little sketch:

![image](https://github.com/mixxxdj/mixxx/assets/57069381/0cf4a0a8-501a-425e-b726-07f2bc0a1e1e)

Swiftb0y replied:
I'd somehow try to put it in the same column. but I'm not a UI designer. Wdyt @ronso0?

danferns replied:
I think it looks more compact:

![image](https://github.com/mixxxdj/mixxx/assets/57069381/d2d9fb2f-6384-4839-8a39-a9eb7c7b1ace)

Swiftb0y replied:
maybe swap the order of the text and the check box so its all still right aligned? idk, IIANA Designer.

Swiftb0y commented:
> I'm concerned about the method used for finding the track's key. Right now, I'm using KeyUtils::guessKeyFromText on the Key string, but I don't think that'll work if the user sets custom Key Notation.

Probably not. You should store the underlying `mixxx::track::io::key::ChromaticKey` in the cell when its text gets set. You can use a custom [item role](https://doc.qt.io/qt-6/model-view-programming.html#item-roles) for that (I think, I'm not much into Qt MV programming tbh).

danferns commented:
Thank you so much @Swiftb0y for the feedback! Preferences for key color palettes is definitely in the plans. I will apply the requested changes and try the item role method.

ronso0 commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+    const int rectWidth = 10; // 2px left padding + 4px width + 4px right padding
+
+    QString elidedText = option.fontMetrics.elidedText(
+            index.data().toString(),
```
```suggestion
            keyText,
```

ronso0 commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+        const auto rgbColor =
+                mixxx::PredefinedColorPalettes::kDefaultKeyColorPalette.at(
+                        openKeyNumber - 1); // Open Key numbers start from 1
+        const auto qcolor = mixxx::RgbColor::toQColor(rgbColor);
```
Just a note: we could move the QString -> QColor translation to KeyUtils so it's available in WKey, too (or some other widget).

ronso0 commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+            index.data().toString(),
+            Qt::ElideLeft,
+            columnWidth(index) - rectWidth);
+
```
Like in LocationDelegate we need to get the highlight color from the palette in order to paint the selected key text correctly:
https://github.com/mixxxdj/mixxx/blob/a48562ad96021bd9571c8fbc8d331291bc30c1fc/src/library/tabledelegates/locationdelegate.cpp#L11-L23

danferns replied:
Thank you! Just fixed this.

ronso0 reviewed (dismissed):
Thank you, this looks nice!

daschuer commented on src/library/basetracktablemodel.cpp:
```diff
@@ -121,7 +122,7 @@ void BaseTrackTableModel::setApplyPlayedTrackColor(bool apply) {
 }
 
-//static
+// static
```
No real issue here in this case. In general we try to avoid reformating untouched code to keep the change set minimal. Make sure your IDE does not automatically reformat whole files.

danferns replied:
Thank you for noting that, I've now set it to only format changes.

daschuer commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,61 @@
+        const QModelIndex& index) const {
+    paintItemBackground(painter, option, index);
+
+    QString keyText = index.data().toString();
```
Here you should request the color directly from the model using a role, the get around the guessing below.

Swiftb0y replied:
yes, guessing the key from the text is inefficient and slow.

ronso0 replied:
Ah, okay. So just get the raw value (key code) via index->data(), cast to int and get the color from `KeyUtils::colorFromNumericValue(int)`?

ronso0 replied:
= don't get the **color** from the model, as requested, just the value

Swiftb0y replied:
yeah, though I wouldn't even go over the `int` index and just store and retrieve the `mixxx::track::io::key::ChromaticKey` directly. otherwise we'll have to deal with out-of-bound values and so forth again.

danferns committed:
refactor: cleanup #includes and move KeyDelegate constructor into header

danferns committed:
fix: use selected text color in KeyDelegate

danferns committed:
move keyToColor translation into KeyUtils

also check bounds for Open Key number inside it.

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>
Co-authored-by: Daniel Schürmann <daschuer@mixxx.org>

daschuer commented on src/track/keyutils.cpp:
```diff
@@ -462,6 +464,20 @@ double KeyUtils::keyToNumericValue(ChromaticKey key) {
+        const auto rgbColor = palette.at(openKeyNumber - 1); // Open Key numbers start from 1
+        return mixxx::RgbColor::toQColor(rgbColor);
+    } else {
+        return QColor::QColor(); // return invalid color
```
```suggestion
        return {}; // return invalid color
```

daschuer commented on src/track/keyutils.cpp:
```diff
@@ -462,6 +464,20 @@ double KeyUtils::keyToNumericValue(ChromaticKey key) {
+
+    if (openKeyNumber != 0) {
+        const auto& palette = mixxx::PredefinedColorPalettes::kDefaultKeyColorPalette;
+        DEBUG_ASSERT(openKeyNumber <= palette.length());
```
```suggestion
        DEBUG_ASSERT(openKeyNumber <= palette.size() && openKeyNumber >= 1);
```

ronso0 commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,56 @@
+    if (keyColor.isValid()) {
+        // Draw the colored rectangle next to the key label
+        painter->fillRect(
+                option.rect.x() + 2,
```
I suggest to remove the left padding in order to allow narrow column configurations.
Would look like this
![image](https://github.com/mixxxdj/mixxx/assets/5934199/dfd8973d-1baa-4def-8ad9-647125366a3b)

danferns replied:
Done. Thank you for this suggestion!

danferns committed:
style: remove left padding of key color bars

This will allow for narrow column setups.

danferns committed:
refactor: pass KEY_ID to KeyDelegate, instead of Key String

Co-authored-by: ronso0 <ronso0@mixxx.org>

danferns committed:
add checkbox to toggle key colors

Add a `KeyColorsEnabled` setting under the `[Config]` group. And render a checkbox for it in the Color preferences.

daschuer commented:
Can you update the screenshot in the initial description of this PR? This way it is more significant if one is later following links from the Changelog.

daschuer commented:
Is the draft state still correct or is it ready for final review and tests?

daschuer commented on src/util/color/predefinedcolorpalettes.cpp:
```diff
@@ -144,6 +144,21 @@ constexpr mixxx::RgbColor kVirtualDJTrackColorBlue(0x0000FF);
 constexpr mixxx::RgbColor kVirtualDJTrackColorFuchsia(0xFF00FF);
 constexpr mixxx::RgbColor kVirtualDJTrackColorWhite(0xFFFFFF);
 
+// Default Mixxx Key Color Palette
```
https://www-datylon-com.cdn.ampproject.org/v/s/www.datylon.com/blog/data-visualization-for-colorblind-readers?amp_gsa=1&amp_js_v=a9&hs_amp=true&usqp=mq331AQIUAKwASCAAgM%3D#amp_tf=Von%20%251%24s&aoh=17198172132733&csi=1&referrer=https%3A%2F%2Fwww.google.com&ampshare=https%3A%2F%2Fwww.datylon.com%2Fblog%2Fdata-visualization-for-colorblind-readers

Interesting are the hue curves of different color blind people. So I guess in a follow up PR, we may offer the full hue curve and reduced versions for color blind people that is only using half of the hue range.

Swiftb0y replied:
We could provide preset palettes optimized for the different types of colorblindness. I think gradients for those already exist online. This would be fairly straightforward to implement once the requusted custom keypalettes are in place.

daschuer reviewed (approved):
Just a minor comment. Rest LGTM, thank you.

danferns commented:
> Is the draft state still correct or is it ready for final review and tests?

Yes, I am not done yet. I still have to read the Key Colors setting and only render the rectangle if it is enabled. 

Also I'd like to squash a couple of commits before review, at least the one which was failing to build.

danferns commented:
I want to read the setting using `ConfigKey("[Config]", "KeyColorsEnabled")` inside `KeyUtils::keyToColor` (if key colors are disabled, return an invalid color). But to do that, I would need access to the [`pConfig`](https://github.com/mixxxdj/mixxx/wiki/developer_guide_config#getting-and-setting-preference-options) variable inside keyutils.cpp. 

I see that the `pConfig` is passed around in class constructors. Is there a way to access it directly?

Should I go about this in some other way entirely?

Swiftb0y commented:
I don't think accessing the config in the KeyUtils is a good idea. Can you share a little more insight on what you're trying to accomplish? Then we can brainstorm together.

Swiftb0y commented:
Thanks for the explanation.
I think that check should stay inside `KeyDelegate`. The static data in `KeyUtils` is ugly and its usefulness is questionable IMO.

danferns commented:
Got it, thank you!

danferns committed:
feat: toggle key colors from preferences

daschuer commented:
In other cases like that we have added a ControlObject 
With the same name. Than you can connect to the value change signal inside the library view and refresh the whole view. I think this is later also required when changing the color schema. So you may consider to add a "None" shema for disable colors.

daschuer commented:
Like here:
https://github.com/mixxxdj/mixxx/blob/7b018fbe588f9c43270e63393179f5cc5cce9e6c/src/preferences/dialog/dlgprefkey.cpp#L48

ronso0 commented:
> I was planning on calling `slotRefreshAllRows()` in the new [`BaseTrackTableModel::setKeyColorsEnabled`](https://github.com/mixxxdj/mixxx/blob/c0657253e34f141f17a37740ef20e4e8b2790b65/src/library/basetracktablemodel.cpp#L122) method, but then realized that this method needs to be static, while `slotRefreshAllRows` is not static.

I'd recommend to avoid that method. That refreshes the entire view and resorts, i.e. like purging tracks this has the undesired resorting side effect, see #12565 

I see that both the new color state and s_bpmColumnPrecision are applied in the next paint event, which happens for me when I close the preferences and see the table being redrawn. Don't know if that happens when the main window is raised again or only if the preferences covered the tracks table.

ronso0 commented:
Btw the key color default is ON, but the checkbox is not checked when opening the Color pref page.

ronso0 commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -77,6 +84,7 @@ DlgPrefColors::~DlgPrefColors() {
 void DlgPrefColors::slotUpdate() {
     comboBoxHotcueColors->clear();
     comboBoxTrackColors->clear();
+    bKeyColorsEnabled->setChecked(m_bKeyColorsEnabled);
```
slotUpdate() is supposed to read the config values and update the view accordingly.
Here, the previous (GUI) value is used (that's why the box is still unchecked even though the setting is ON), and it's not obvious that `m_colorPaletteSettings.get..` queries the config.
Please read the config with `m_pConfig->getValue([color setting])`.

ronso0 commented:
FWIW I expected the key color setting in the Library preferences since that is what handle the settings and appearance of the tracks table.

danferns committed:
fix: checkbox ui not in sync with key colors setting

danferns committed:
refactor: remove redundant canConvert assertion

danferns committed:
mention source of Mixxx Key Color Palette colors

ronso0 commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -345,6 +358,13 @@ void DlgPrefColors::slotEditHotcuePaletteClicked() {
+    m_bKeyColorsEnabled = static_cast<bool>(i);
+    BaseTrackTableModel::setKeyColorsEnabled(m_bKeyColorsEnabled);
+    m_pConfig->setValue(kKeyColorsEnabledConfigKey, bKeyColorsEnabled->checkState());
+    slotUpdate();
```
This slot is called when the checkbox is clicked (changed), so the checkbox state and the config state are now identical.
No need to update the entire page now.

ronso0 replied:
Sorry for the fragmented review, I didn't take a closer look earlier.

danferns replied:
I did not quite get you. Did you mean that we don't need the `slotUpdate()` call because it gets called some other way when the checkbox is clicked?

ronso0 replied:
Yes, we don't need it because the checkbox state and the config state are identical after writing to the config and `setKeyColorsEnabled()`.

danferns committed:
fix indentation and set colspan in Color Preferences UI

danferns commented on src/preferences/dialog/dlgprefcolorsdlg.ui:
```diff
@@ -100,7 +100,19 @@
+      <string>Enable Key Colors</string>
+     </property>
+    </widget>
+   </item>
```
Fixed, thank you.

ronso0 commented:
> FWIW I expected the key color setting in the Library preferences since that is what handle the settings and appearance of the tracks table.

What do you think?
Should we have the enable switch in the Library page, and the future palette config in Colors?

danferns commented:
> What do you think? Should we have the enable switch in the Library page, and the future palette config in Colors?

If we add key color support to the WKey widget in the future, then it feels more natural to me to have this setting under the Color preferences. Because then this toggle is not only changing the library view but also the deck view.

danferns committed:
remove unnecessary slotUpdate() call

danferns commented:
Thank you @daschuer, @ronso0 and @Swiftb0y for your feedback so far! I think I'm ready for a final review.

I was thinking of cleaning up the git log a little, but I think I'll do it at the end before merging (would that be fine?). There might still be more changes to be made from now, and I'm also not sure how much to clean up.

Swiftb0y commented:
Yes, don't worry with cleaning up the log for now. I'm sure there will be a bunch more comments.

danferns commented:
I want to work on adding the key color palettes. I'm a bit unsure which branch to do it on. Should I branch off from this PR's branch, and then create a separate PR for the palettes? Or would it be better to keep working on this branch only?

ronso0 commented:
I suggest to make "small" steps, i.e. get this merged first, then improve / tweak / extend this new feature in separate branches.

danferns commented:
Got it, thank you.

ronso0 reviewed (commented):
Alrighty, this LGTM except a few bits.

ronso0 commented:
> maybe swap the order of the text and the check box so its all still right aligned? idk, IIANA Designer.

IMHO it's okay like this.
IIRC it's not possible to swap checkbox and label, that would need to be two separate widgets (QLabel + QCheckBox without label).

danferns committed:
refactor: set keyColor, keyText as const

Co-authored-by: ronso0 <ronso0@mixxx.org>

danferns committed:
refactor: remove KeyUtils scope used in keyToColor

Co-authored-by: ronso0 <ronso0@mixxx.org>

danferns committed:
refactor: update key colors checkbox name

also add checkbox to <tabstops>

Co-authored-by: ronso0 <ronso0@mixxx.org>

danferns committed:
clear keyText when key is invalid

ronso0 commented on src/library/tabledelegates/keydelegate.cpp:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    QString keyText = KeyUtils::keyToString(key);
+    if (QString("INVALID").compare(keyText) == 0) { // if the key is invalid, display empty string
```
not the other way around, since we already have a QString and don't need to construct another one?
(based what I think I understood fo QString stuff)
```suggestion
    if (keyText.compare("INVALID") == 0) { // if the key is invalid, display empty string
```

Swiftb0y replied:
good suggestion, but better to avoid string comparison altogether. https://github.com/mixxxdj/mixxx/pull/13390#discussion_r1672347393

danferns committed:
refactor: draw key text/color only if key is valid

danferns commented:
Thank you for the review. I have used the enum comparison now, and placed all the text/color rendering code inside this condition. 

Let me know when it is the right time for rebasing.

ronso0 reviewed (approved):
Yep, LGTM : )

Swiftb0y commented on src/preferences/colorpalettesettings.h:
```diff
@@ -3,6 +3,8 @@
 #include "preferences/usersettings.h"
 #include "util/color/colorpalette.h"
 
+#define KEY_COLORS_ENABLED "KeyColorsEnabled"
```
can you remove the `#define`? just copy paste it into its only use below.

Swiftb0y reviewed (commented):
not perfect, but good enough considering the status quo. almost LGTM.

danferns committed:
remove #define KEY_COLORS_ENABLED

danferns commented:
> not perfect, but good enough considering the status quo.

I appreciate the honest feedback. Can you please tell me in what aspects I can improve?

Swiftb0y commented:
> I appreciate the honest feedback. 

Don't get me wrong, this PR is quite great already ;). There is just always room to improve. 

> Can you please tell me in what aspects I can improve?

The primary issue I have is that you followed some suboptimal patterns that are already in the existing code. I'm not expecting you to clean those up, though I'm not particularly fond of the fact that this PR is adding more suboptimal code. Refactoring the code in question is probably outside your skillset and GSoC project scope but there are a bunch of refactorings that are in order IMO. Ideally, all the static global variables (the ones prefixed `s_`) should be eliminated (not sure how, haven't looked too deep into it). `src/util/color/predefinedcolorpalettes.cpp` could use a little bit of a cleanup. `KeyDelegate::paintItem` could use a little more reordering to simplify its controlflow (prefer early returns) and separate the individual steps better (currently a lot of the low-level math of the drawing code is interspersed with the high-level drawing code) (see [SLAP](http://principles-wiki.net/principles:single_level_of_abstraction)). Passing `s_keyColorsEnabled` as part of the `BaseTrackTableModel` is weird (especially via the `roleValue()`) because the value doesn't change depending on the index.

Having said all that, I'm not expecting you to address these concerns. Instead discuss with your mentor (@daschuer IIRC?) on how to proceed.

Swiftb0y reviewed (approved):
LGTM. Thank you. Who wants to press the button? Did you want to rebase first?

danferns commented:
Thank you so much, this is very helpful. I'll discuss with Daniel for the next steps, and I'll keep these points in mind for future PRs. I am growing a lot thanks to this project. :) 

I do want to rebase!

daschuer commented:
Thank you for the review. Yes, there is room for improvements. On the other hand we have tight plan for the remaining project. From this point of view I would rather see a beautiful Heat column instead of beautiful code. 

We have already the green check marks so I will merge once @danferns gives a go. 

If you feeling lucky to address some of @Swiftb0y points, go ahead and propose it in a separate PR.

danferns commented:
I'm done with the rebase! I squashed only a build error, and two commits involving indentation / comments. If anything else is needed then let me know, but otherwise I am ready for this PR to be merged!

Swiftb0y commented:
> On the other hand we have tight plan for the remaining project. From this point of view I would rather see a beautiful Heat column instead of beautiful code.

I agree in this case.

daschuer commented:
Thank you very much! This looks really nice and works good.

acolombier commented:
After synching my dev branch, I just seen that eye candy appeared. It looks very good! Thanks for that feature!

Eve00000 commented:
I tested this today, some remarks:
- I don't have a key palette under 'Enable Key Colors' in Preferences -> Colors
-Pushing 'Apply' after changing 'Enable Key Colors' does not work, changes only happen after push on 'OK'
- Nice color bar in 'Key'-column but Key text is cut of on the left side (normally text is cut of on the right side) while my most important part is on the left side.
![afbeelding](https://github.com/user-attachments/assets/98bd8d89-07db-46a0-8e88-71588e9f1a68)
![afbeelding](https://github.com/user-attachments/assets/a718ce5b-d819-402d-b200-b3535a599216)

ronso0 commented:
> * I don't have a key palette under 'Enable Key Colors' in Preferences -> Colors

There is only one palette atm.

Repaint with/without colors only happens when the table is repainted after the preferences are closed.
